C++/126_Word_Ladder_II.cpp

```diff
@@ -0,0 +1,330 @@
+// 126. Word Ladder II
+/**
+ * Given two words (beginWord and endWord), and a dictionary's word list, find all shortest transformation sequence(s) from beginWord to endWord, such that:
+ * 
+ * 1. Only one letter can be changed at a time
+ * 2. Each intermediate word must exist in the word list
+ * 
+ * For example,
+ * Given:
+ * beginWord = "hit"
+ * endWord = "cog"
+ * wordList = ["hot","dot","dog","lot","log"]
+ *
+ * Return
+ *   [
+ *     ["hit","hot","dot","dog","cog"],
+ *     ["hit","hot","lot","log","cog"]
+ *   ]
+ * 
+ * Note:
+ * All words have the same length.
+ * All words contain only lowercase alphabetic characters.
+ * 
+ * Tags: Array, Backtracking, Breadth-first Search, String
+ * 
+ * Author: Kuang Qin
+ */
+
+#include "stdafx.h"
+#include <vector>
+#include <string>
+#include <unordered_set>
+#include <unordered_map>
+#include <queue>
+#include <iostream>
+
+using namespace std;
+
+// BFS + DFS
+class Solution {
+    // find all possible transformations of word in wordList
+    vector<string> transform(string &word, unordered_set<string> &wordList) {
+        vector<string> transList;
+        for (int i = 0; i < word.size(); i++)
+        {
+            for (char ch = 'a'; ch <= 'z'; ch++)
+            {
+                if (word[i] != ch)    // skip the same word
+                {
+                    string str(word);
+                    str[i] = ch;
+                    if (wordList.find(str) != wordList.end())
+                    {
+                        // if the transformed word is in wordList, add to result
+                        transList.push_back(str);
+                    }
+                }
+            }
+        }
+
+        return transList;
+    }
+
+    // create a queue to track the distance of each word in wordList from beginWord
+    void bfs(unordered_map<string, vector<string>> &transMap, unordered_map<string, int> &distMap, 
+                string &beginWord, string &endWord, unordered_set<string> &wordList) {
+        queue<string> q;
+        q.push(beginWord);
+
+        // make sure beginWord and endWord is in wordList
+        wordList.insert(beginWord);
+        wordList.insert(endWord);
+
+        // start from the beginWord
+        distMap[beginWord] = 0;
+
+        while (!q.empty())
+        {
+            int count = q.size();
+            bool isEndFound = false;    // flag if found the endWord
+            for (int i = 0; i < count; i++)    // bfs by level
+            {
+                string word = q.front();
+                q.pop();
+
+                vector<string> transList = transform(word, wordList);
+                vector<string> *curList = &transMap[word];
+                for (int j = 0; j < transList.size(); j++)
+                {
+                    string str(transList[j]);
+
+                    // update possible transformations for each word in transList
+                    curList->push_back(str);
```

Comment:
Not sure. And runtime doesn't really matter as long as you understand the complexity and the logic.